robolectric/src/test/java/org/robolectric/shadows/ShadowBiometricManagerTest.java

```diff
@@ -21,22 +25,78 @@ public class ShadowBiometricManagerTest {
   @Before
   public void setUp() {
     biometricManager =
-            ApplicationProvider.getApplicationContext().getSystemService(BiometricManager.class);
+        ApplicationProvider.getApplicationContext().getSystemService(BiometricManager.class);
     assertThat(biometricManager).isNotNull();
   }
 
   @Test
   public void testCanAuthenticate_serviceNotConnected_canNotAuthenticate() {
-    ((ShadowBiometricManager) Shadow.extract(biometricManager)).setCanAuthenticate(false);
+    ShadowBiometricManager shadowBiometricManager = Shadow.extract(biometricManager);
+    shadowBiometricManager.setCanAuthenticate(false);
 
     assertThat(biometricManager.canAuthenticate())
         .isEqualTo(BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE);
   }
 
   @Test
   public void testCanAuthenticate_serviceConnected_canAuthenticate() {
-    ((ShadowBiometricManager) Shadow.extract(biometricManager)).setCanAuthenticate(true);
+    ShadowBiometricManager shadowBiometricManager = Shadow.extract(biometricManager);
+    shadowBiometricManager.setCanAuthenticate(true);
 
     assertThat(biometricManager.canAuthenticate()).isEqualTo(BiometricManager.BIOMETRIC_SUCCESS);
   }
+
+  @Test
+  public void testCanAuthenticate_serviceNotConnected_noEnrolledBiometric_biometricNotEnrolled() {
+    ShadowBiometricManager shadowBiometricManager = Shadow.extract(biometricManager);
+    shadowBiometricManager.setEnrolledBiometrics(0);
+
+    assertThat(biometricManager.canAuthenticate())
+        .isEqualTo(BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED);
+  }
+
+  @Test
+  public void testCanAuthenticate_serviceNotConnected_noHardware_biometricHwUnavailable() {
+    ShadowBiometricManager shadowBiometricManager = Shadow.extract(biometricManager);
+    shadowBiometricManager.hasHardware(false);
+
+    assertThat(biometricManager.canAuthenticate())
+        .isEqualTo(BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE);
+  }
+
+  @Test
+  @Config(minSdk = R)
+  public void
+      testCanAuthenticate_serviceNotConnected_securityUpdateRequired_biometricErrorSecurityUpdateRequired() {
+    ShadowBiometricManager shadowBiometricManager = Shadow.extract(biometricManager);
+    shadowBiometricManager.requireSecurityUpdate(true);
+
+    assertThat(biometricManager.canAuthenticate())
+        .isEqualTo(BiometricManager.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED);
+  }
+
+  @Test
+  @Config(minSdk = R)
+  public void testCanAuthenticateBiometricWeak_NoWeakButHaveStrongEnrolled_canAuthenticate() {
```

Comment:
Rename to `testCanAuthenticateBiometricWeak_serviceConnected_noWeakButHaveStrongEntrolled_canAuthenticate`?

Reply:
Sure no problem